js/lib/BTCUtils.js

```diff
@@ -0,0 +1,659 @@
+   */
+  extractPrevBlockLE: (header) => {
+    return utils.safeSlice(header, 4, 36);
+  },
+
+  /**
+   *  @notice               Extracts the previous block's hash from a block header
+   * @dev                   Block headers do NOT include block number :(
+   * @param {Uint8Array}    header The header
+   * @returns {Uint8Array}  The previous block's hash (big-endian)
+   */
+  extractPrevBlockBE: (header) => {
+    return module.exports.reverseEndianness(module.exports.extractPrevBlockLE(header));
+  },
+
+  /**
+   * @notice                Extracts the timestamp from a block header
+   * @dev                   Time is not 100% reliable
+   * @param {Uint8Array}    header The header
+   * @returns {Uint8Array}  The timestamp (little-endian bytes)
+   */
+  extractTimestampLE: (header) => {
+    return utils.safeSlice(header, 68, 72);
+  },
+
+  /**
+   * @notice                Extracts the timestamp from a block header
+   * @dev                   Time is not 100% reliable
+   * @param {Uint8Array}    header The header
+   * @returns {BigInt}      The timestamp (uint)
+   */
+  extractTimestamp: (header) => {
+    return utils.bytesToUint(module.exports.reverseEndianness(module.exports.extractTimestampLE(header)));
+  },
+
+  /**
+   * @notice                Extracts the expected difficulty from a block header
+   * @dev                   Does NOT verify the work
+   * @param {Uint8Array}    header The header
+   * @returns {BigInt}      The difficulty as an integer
+   */
+  extractDifficulty: (header) => {
+    return module.exports.calculateDifficulty(module.exports.extractTarget(header));
+  },
+
+  /**
+   * @notice                Concatenates and hashes two inputs for merkle proving
+   * @param {Uint8Array}    a The first hash
+   * @param {Uint8Array}    b The second hash
+   * @returns {Uint8Array}  The double-sha256 of the concatenated hashes
+   */
+  hash256MerkleStep: (a, b) => {
+    return module.exports.hash256(utils.concatUint8Arrays([a, b]));
+  },
+
+  /**
+   * @notice                Verifies a Bitcoin-style merkle tree
+   * @dev                   Leaves are 1-indexed.
+   * @param {Uin8Array}     proof The proof. Tightly packed LE sha256 hashes. The last hash is the root
+   * @param {Number}        index The index of the leaf
+   * @returns {Boolean}     True if the proof is value, else false
+   */
+  verifyHash256Merkle: (proof, index) => {
+    const proofLength = proof.length;
+
+    // Not an even number of hashes
+    if (proofLength % 32 !== 0) {
+      return false;
+    }
+
+    // Special case for coinbase-only blocks
+    if (proofLength === 32) {
+      return true;
+    }
+
+    // Should never occur
+    if (proofLength === 64) {
+      return false;
+    }
+
+    let idx = BigInt(index);
+    let root = utils.safeSlice(proof, (proofLength - 32), proofLength);
+    let current = utils.safeSlice(proof, 0, 32);
+
+    for (let i = 1; i < ((proofLength / 32) - 1); i++) {
```

Comment:
let's add a line here:
```
let next = utils.safeSlice(proof, (i * 32), ((i * 32) + 32))
```

then we can make the following block:

```
if (idx % 2n === 1n) {
        current = module.exports.hash256MerkleStep(next, current);
      } else {
        current = module.exports.hash256MerkleStep(current, next);
      }
```

Reply:
smart, fixed